src/arraysense/store/rollup.py

```
@@ -619,12 +642,31 @@ def rebuild_circuit_hourly(conn: sqlite3.Connection, start: int, end: int) -> No
     part = _floor_div("timestamp", 3600)
+    # The window function partitions by the bucket as well as the circuit, so
+    # each hour is measured from its own readings alone and a rebuild gives the
+    # same answer whatever range it was asked for. A reading with nothing after
+    # it in its hour takes the full interval.
+    spans = (
+        f"SELECT {part} * 3600 AS bucket, circuit_id, watts, "
+        "MIN(COALESCE("
+        f"  LEAD(timestamp) OVER (PARTITION BY circuit_id, {part} ORDER BY timestamp) - timestamp,"
+        "  ?"
+        "), ?) AS covered "
+        "FROM circuit_reading WHERE timestamp >= ? AND timestamp < ?"
```

Ruff (0.16.1)

[error] Possible SQL injection vector through string-based query construction

(S608)